compiler/rustc_middle/src/ty/query.rs

```diff
@@ -106,6 +106,55 @@ impl<'tcx> TyCtxt<'tcx> {
     }
 }
 
+#[inline(always)]
+fn query_get_at<'tcx, Cache, K>(
+    tcx: TyCtxt<'tcx>,
+    execute_query: fn(
+        &'tcx dyn QueryEngine<'tcx>,
+        TyCtxt<'tcx>,
+        Span,
+        Cache::Key,
+        QueryMode,
+    ) -> Option<Cache::Stored>,
+    query_cache: &Cache,
+    span: Span,
+    key: K,
+) -> Cache::Stored
+where
+    K: IntoQueryParam<Cache::Key>,
+    Cache::Stored: Copy,
+    Cache: QueryCache,
+{
+    let key = key.into_query_param();
+    match try_get_cached(tcx, query_cache, &key) {
+        Some(value) => value,
+        None => execute_query(tcx.queries, tcx, span, key, QueryMode::Get).unwrap(),
+    }
+}
+
+#[inline(always)]
+fn query_ensure<'tcx, Cache, K>(
```

Comment:
Can this be merged with `query_get_at` by adding a `QueryMode` parameter and returning an `Option`?

Reply:
This change was a regression:
<table><tr><td rowspan="2">Benchmark</td><td colspan="1"><b>Before</b></th><td colspan="2"><b>After</b></th></tr><tr><td align="right">Time</td><td align="right">Time</td><td align="right">%</th></tr><tr><td>🟣 <b>clap</b>:check</td><td align="right">1.7607s</td><td align="right">1.7700s</td><td align="right"> 0.52%</td></tr><tr><td>🟣 <b>hyper</b>:check</td><td align="right">0.2565s</td><td align="right">0.2576s</td><td align="right"> 0.44%</td></tr><tr><td>🟣 <b>regex</b>:check</td><td align="right">0.9807s</td><td align="right">0.9862s</td><td align="right"> 0.56%</td></tr><tr><td>🟣 <b>syn</b>:check</td><td align="right">1.5808s</td><td align="right">1.5889s</td><td align="right"> 0.51%</td></tr><tr><td>🟣 <b>syntex_syntax</b>:check</td><td align="right">6.1378s</td><td align="right">6.1567s</td><td align="right"> 0.31%</td></tr><tr><td>Total</td><td align="right">10.7165s</td><td align="right">10.7594s</td><td align="right"> 0.40%</td></tr><tr><td>Summary</td><td align="right">1.0000s</td><td align="right">1.0047s</td><td align="right"> 0.47%</td></tr></table>